File: tools/glass-pane/frontend/src/types/events.ts
/**
 * Server-Sent Events (SSE) types for Glass Pane frontend.
 */

import { Log } from './api'

export type SSEEventType =
  | 'phase_update'
  | 'file_created'
  | 'log_batch'
  | 'metrics_update'
  | 'job_status_change'
  | 'heartbeat'

export interface PhaseUpdateData {
  phase: string
  status: string
  description: string
}

export interface FileCreatedData {
  path: string
  timestamp: string
}

export interface LogBatchData {
  logs: Log[]
}

export interface MetricsUpdateData {
  tokens_used: number
  duration: number
  files: number
}

export interface JobStatusChangeData {
  status: 'running' | 'completed' | 'failed'
}

export interface HeartbeatData {
  timestamp: string
}

export type SSEEventData =
  | PhaseUpdateData
  | FileCreatedData
  | LogBatchData
  | MetricsUpdateData
  | JobStatusChangeData
  | HeartbeatData

export interface SSEEvent {
  type: SSEEventType
  data: SSEEventData
}

export interface SSEMessage {
  event: string
  data: string
}

File: tools/glass-pane/backend/models/events.py
"""
Server-Sent Events (SSE) data models.
"""

from pydantic import BaseModel
from typing import Literal


class PhaseUpdateEvent(BaseModel):
    """Phase change event."""

    type: Literal["phase_update"] = "phase_update"
    data: dict  # {phase: str, status: str, description: str}


class FileCreatedEvent(BaseModel):
    """File creation event."""

    type: Literal["file_created"] = "file_created"
    data: dict  # {path: str, timestamp: str}


class LogBatchEvent(BaseModel):
    """Batch of new log entries."""

    type: Literal["log_batch"] = "log_batch"
    data: dict  # {logs: List[Log]}


class MetricsUpdateEvent(BaseModel):
    """Build metrics update."""

    type: Literal["metrics_update"] = "metrics_update"
    data: dict  # {tokens_used: int, duration: int, files: int}


class JobStatusChangeEvent(BaseModel):
    """Job status change event."""

    type: Literal["job_status_change"] = "job_status_change"
    data: dict  # {status: str}


class HeartbeatEvent(BaseModel):
    """Keep-alive heartbeat."""

    type: Literal["heartbeat"] = "heartbeat"
    data: dict  # {timestamp: str}


# Union type for all SSE events
SSEEvent = (
    PhaseUpdateEvent
    | FileCreatedEvent
    | LogBatchEvent
    | MetricsUpdateEvent
    | JobStatusChangeEvent
    | HeartbeatEvent
)
